grpc_attestation/src/server.rs

```diff
@@ -0,0 +1,203 @@
+//
+// Copyright 2022 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+//! Server-side implementation of the bidirectional streaming gRPC remote attestation handshake
+//! protocol.
+
+use crate::proto::{
+    streaming_session_server::StreamingSession, StreamingRequest, StreamingResponse,
+};
+use anyhow::Context;
+use futures::{Stream, StreamExt};
+use oak_remote_attestation::handshaker::{AttestationBehavior, Encryptor, ServerHandshaker};
+use std::pin::Pin;
+use tonic::{Request, Response, Status, Streaming};
+
+/// Handler for subsequent encrypted requests from the stream after the handshake is completed.
+struct EncryptedRequestHandler<F, S>
+where
+    F: Send + Sync + Clone + FnOnce(Vec<u8>) -> S,
+    S: std::future::Future<Output = anyhow::Result<Vec<u8>>> + Send + Sync,
+{
+    /// The stream of incoming requests.
+    request_stream: Streaming<StreamingRequest>,
```

Comment:
Should this be a generic stream? (not tonic specific). And could the struct also implement stream interface? i.e. could this be expressed as a transformation between streams? e.g. with https://docs.rs/futures/latest/futures/stream/trait.StreamExt.html#method.map